readers/apiserver/watchlist/watchlist.go

```diff
@@ -1,251 +1,639 @@
+	finished := make(chan struct{})
+	timer := time.After(30 * time.Second)
 
 	go func() {
-		if r.r.Relist() <= 0 {
-			return
+		select {
+		case <-finished:
+		case <-timer:
+			cancel()
 		}
+	}()
+
+	var spanWatchers []spawnWatcher
+	switch rt {
+	case types.RTNamespace:
+		spanWatchers = r.createNamespaceWatcher(ctx)
+	case types.RTNode:
+		spanWatchers = r.createNodesWatcher(ctx)
+	case types.RTPod:
+		spanWatchers = r.createPodsWatcher(ctx)
+	case types.RTPersistentVolume:
+		spanWatchers = r.createPersistentVolumesWatcher(ctx)
+	case types.RTRBAC:
+		spanWatchers = r.createRBACWatcher(ctx)
+	case types.RTService:
+		spanWatchers = r.createServicesWatcher(ctx)
+	case types.RTDeployment:
+		spanWatchers = r.createDeploymentsWatcher(ctx)
+	case types.RTIngressController:
+		spanWatchers = r.createIngressesWatcher(ctx)
+	case types.RTEndpoint:
+		spanWatchers = r.createEndpointsWatcher(ctx)
+	default:
+		return fmt.Errorf("unknown resource type: %v", rt)
+	}
 
-		t := time.NewTimer(r.r.Relist())
-		defer t.Stop()
+	if err := r.watch(ctx, rt, spanWatchers); err != nil {
+		return fmt.Errorf("error starting %s watcher: %v", rt, err)
+	}
+	if ctx.Err() != nil {
+		return fmt.Errorf("error starting namespace watcher: could not connect to server in time")
+	}
+	close(finished)
+	return ctx.Err()
+}
 
-		for {
-			if exit := r.switchWait(t); exit {
-				return
+func (r *Reader) createNamespaceWatcher(ctx context.Context) []spawnWatcher {
+	return []spawnWatcher{
+		func(options metav1.ListOptions) (watch.Interface, error) {
+			wi, err := r.clientset.CoreV1().Namespaces().Watch(ctx, options)
+			if err != nil {
+				return nil, err
 			}
-		}
-	}()
+			return wi, nil
+		},
+	}
 }
 
-// switchWait will wait for the timer to expire or the close channel to close.
-// If the close channel closes, this function will return true.
-// If the timer expires, this function will switch the client and reset the timer.
-func (r *Reader) switchWait(t *time.Timer) (exit bool) {
-	select {
-	case <-r.closeCh:
-		return true
-	case <-t.C:
-		r.clientSwitchRetry()
-		resetTimer(r.r.Relist(), t)
-		return false
+func (r *Reader) createNodesWatcher(ctx context.Context) []spawnWatcher {
+	return []spawnWatcher{
+		func(options metav1.ListOptions) (watch.Interface, error) {
+			wi, err := r.clientset.CoreV1().Nodes().Watch(ctx, options)
+			if err != nil {
+				return nil, err
+			}
+			return wi, nil
+		},
 	}
 }
 
-// clientSwitchRetry will create a new reader object and close the old one. If there is
-// an error creating the new reader, it will retry until it is successful or Close() is called.
-func (r *Reader) clientSwitchRetry() {
-	if r.testClientSwitchRetry != nil {
-		r.testClientSwitchRetry()
-		return
+func (r *Reader) createPodsWatcher(ctx context.Context) []spawnWatcher {
+	return []spawnWatcher{
+		func(options metav1.ListOptions) (watch.Interface, error) {
+			wi, err := r.clientset.CoreV1().Pods("").Watch(ctx, options)
```

Comment:
do we want to consistently use "" or metav1.NamespaceAll for these watchers

Reply:
Honestly, it doesn't matter.  Its a solution to a flawed design where "All".  However, I've converted all references everywhere for pods, services, .... to metav1.NamespaceAll .